File: GeoJsonOracle.java
package pzformat;

import java.nio.file.*;
import java.util.*;

/**
 * Cross-language oracle for Json + GeoJson, Java side.
 *
 *   emit <input.geojson> <out.txt>
 *
 * Emits a canonical digest of the parsed feature list for cross-language diff,
 * in the ClassifyOracle style: one tab-separated record per line, fixed field
 * order, "-" for absent.
 *
 * TWO DEPARTURES from ClassifyOracle, both deliberate:
 *
 *  1. NOT SORTED. ClassifyOracle sorts because tile names are a set. Here the
 *     feature order IS the contract — GisImport.buildings is "in import order"
 *     and per-building RNG seeding indexes into that list, so sorting would
 *     hide exactly the class of bug this oracle exists to catch. Emitting in
 *     file order also sidesteps the Java-UTF-16 vs C++-UTF-8 collation
 *     difference entirely rather than having to defend against it.
 *
 *  2. COORDINATES AS RAW BITS. Doubles are emitted as %016x of their bit
 *     pattern, not as text. Comparing formatted coordinates would test the two
 *     printf implementations rather than the two parsers; comparing bits tests
 *     that the parse produced the identical double. Property values still go
 *     through asText(), because there the string conversion IS the behaviour
 *     under test.
 */
public final class GeoJsonOracle {

    /** Escape control characters so one record is always exactly one line with
     *  fixed tab positions. Without this a property value containing a tab or a
     *  newline (OSM description tags do) silently misaligns the digest fields —
     *  identically on both sides, so the diff would still pass while comparing
     *  the wrong columns. */
    static String esc(String s) {
        if (s == null) return "-";
        StringBuilder b = new StringBuilder(s.length());
        for (int i = 0; i < s.length(); i++) {
            char c = s.charAt(i);
            switch (c) {
                case '\\' -> b.append("\\\\");
                case '\t'  -> b.append("\\t");
                case '\n'  -> b.append("\\n");
                case '\r'  -> b.append("\\r");
                default -> {
                    if (c < 0x20) b.append(String.format("\\x%02x", (int) c));
                    else b.append(c);
                }
            }
        }
        return b.toString();
    }

    public static void main(String[] args) throws Exception {
        Path in  = Path.of(args[0]);
        Path out = Path.of(args[1]);

        GeoJson g = GeoJson.read(in);
        List<String> lines = new ArrayList<>();

        for (int i = 0; i < g.features.size(); i++) {
            GeoJson.Feature f = g.features.get(i);
            int pts = 0;
            for (List<double[]> r : f.rings) pts += r.size();
            String type = (f.type == null || f.type.isEmpty()) ? "-" : f.type;
            lines.add("F\t" + i + "\t" + esc(type) + "\t" + f.rings.size() + "\t" + pts);

            for (Map.Entry<String, String> e : f.props.entrySet()) {
                String v = e.getValue();
                lines.add("P\t" + i + "\t" + esc(e.getKey()) + "\t" + esc(v));
            }

            for (int r = 0; r < f.rings.size(); r++) {
                List<double[]> ring = f.rings.get(r);
                for (int j = 0; j < ring.size(); j++) {
                    double[] p = ring.get(j);
                    lines.add("C\t" + i + "\t" + r + "\t" + j + "\t"
                            + String.format("%016x", Double.doubleToRawLongBits(p[0])) + "\t"
                            + String.format("%016x", Double.doubleToRawLongBits(p[1])));
                }
            }
        }

        Files.write(out, lines);
        System.out.println("java geojson: " + g.features.size() + " features, "
                + lines.size() + " digest lines -> " + out);
    }
}

File: geojson_oracle.cpp
// geojson_oracle.cpp — C++ side of the Json + GeoJson cross-language oracle.
//
//   geojson_oracle <input.geojson> <out.txt>
//
// Must produce a file byte-identical to GeoJsonOracle.java's output. See that
// file for why the digest is unsorted and why coordinates are emitted as bits.

#include <cstdint>
#include <cstdio>
#include <cstring>
#include <fstream>
#include <string>

#include "geojson.hpp"

using namespace pzformat;

// Escape control characters so one record is always exactly one line with fixed
// tab positions. Must match GeoJsonOracle.java's esc() byte for byte. Operates
// on UTF-8 bytes; every byte of a multi-byte sequence is >= 0x80 so none is
// touched, which is what makes this agree with the Java char-wise version.
static std::string esc(const std::string& s) {
    std::string b;
    b.reserve(s.size());
    for (unsigned char c : s) {
        switch (c) {
            case '\\': b += "\\\\"; break;
            case '\t': b += "\\t"; break;
            case '\n': b += "\\n"; break;
            case '\r': b += "\\r"; break;
            default:
                if (c < 0x20) {
                    char t[8];
                    std::snprintf(t, sizeof t, "\\x%02x", c);
                    b += t;
                } else {
                    b.push_back(static_cast<char>(c));
                }
        }
    }
    return b;
}

static std::string bitsHex(double d) {
    std::uint64_t bits;
    std::memcpy(&bits, &d, sizeof bits);
    char buf[32];
    std::snprintf(buf, sizeof buf, "%016llx", static_cast<unsigned long long>(bits));
    return buf;
}

int main(int argc, char** argv) {
    if (argc < 3) {
        std::fprintf(stderr, "usage: geojson_oracle <input.geojson> <out.txt>\n");
        return 2;
    }

    GeoJson g = GeoJson::read(argv[1]);

    std::ofstream out(argv[2], std::ios::binary);
    std::size_t lines = 0;
    auto emit = [&](const std::string& s) {
        out << s << '\n';
        lines++;
    };

    for (std::size_t i = 0; i < g.features.size(); i++) {
        const auto& f = g.features[i];
        std::size_t pts = 0;
        for (const auto& r : f.rings) pts += r.size();
        const std::string type = f.type.empty() ? "-" : f.type;
        emit("F\t" + std::to_string(i) + "\t" + esc(type) + "\t" + std::to_string(f.rings.size()) +
             "\t" + std::to_string(pts));

        for (const auto& kv : f.props)
            emit("P\t" + std::to_string(i) + "\t" + esc(kv.first) + "\t" + esc(kv.second));

        for (std::size_t r = 0; r < f.rings.size(); r++) {
            const auto& ring = f.rings[r];
            for (std::size_t j = 0; j < ring.size(); j++) {
                emit("C\t" + std::to_string(i) + "\t" + std::to_string(r) + "\t" +
                     std::to_string(j) + "\t" + bitsHex(ring[j].lon) + "\t" +
                     bitsHex(ring[j].lat));
            }
        }
    }

    std::printf("cpp  geojson: %zu features, %zu digest lines -> %s\n", g.features.size(), lines,
                argv[2]);
    return 0;
}
